### DISCUSSION TITLE
[mini.pick] pick "buffer lines" in current buffer and include treesitter syntax highlighting

### DISCUSSION THREAD

**Discussion by abeldekat** [nvim-mini/mini.nvim#988] [Show and tell]
![1719252990](https://github.com/echasnovski/mini.nvim/assets/58370433/fc72a78e-89ba-4c71-b417-6aa2040a6deb)

For the current buffer, I modified the `buf_lines` picker to show the lines with treesitter syntax highlighting.

The [code](https://github.com/abeldekat/nvim_pde/blob/main/lua/ak/config/editor/mini_pick.lua):

```lua
Pick.registry.buffer_lines_current = function()
  local show = function(buf_id, items, query, opts)
    if items and #items > 0 then -- one buffer, one ft: Enable highlighting
      local ft = vim.bo[items[1].bufnr].filetype
      -- vim.bo.filetype = ft -- Safer:
      vim.bo[buf_id].filetype = ft
    end
    Pick.default_show(buf_id, items, query, opts)
  end
  MiniExtra.pickers.buf_lines({ scope = "current" }, { source = { show = show } })
end
```

I use this quite often to fuzzy search a function while the `lsp` is starting up.

**EDIT 20240820**: [Latest version](https://github.com/echasnovski/mini.nvim/discussions/988#discussioncomment-10398788)

**Comment by echasnovski** [comment]
Oh, that's a nice one!

Couple of remarks:
- I think `vim.bo.filetype = ft` is safer to be `vim.bo[buf_id].filetype = ft`.
- Unfortunately, this is not universal, as shown lines start with digits and `:`. I have doubts that every tree-sitter parser handles this well.

**Reply by abeldekat** [reply]
Thanks! I am having a *really* good time with mini.pick....)

> I think vim.bo.filetype = ft is safer to be vim.bo[buf_id].filetype = ft

I need to use the filetype of the buffer being displayed, so I just use the filetype of the first item's bufnr to activate the highlighting in the picker list. Is there another way? I think `vim.bo.filetype` refers to the filetype of the buffer of the picker....

> Unfortunately, this is not universal, as shown lines start with digits and :. I have doubts that every tree-sitter parser handles this well.

Fingers crossed... I am already quite happy it works for `lua`. This behavior is the standard in Telescope's `current_buffer_fuzzy_find`. I also looked at `Fzf-Lua`: No syntax highlighting in `lines` and `blines`.

**Reply by echasnovski** [reply]
> > I think vim.bo.filetype = ft is safer to be vim.bo[buf_id].filetype = ft
> 
> I need to use the filetype of the buffer being displayed, so I just use the filetype of the first item's bufnr to activate the highlighting in the picker list. Is there another way? I think `vim.bo.filetype` refers to the filetype of the buffer of the picker....

`vim.bo.filetype` without index operates on the *current* buffer. The fact that this works is mostly a coincidence.
The first argument with which `source.show()` is called is [identifier of target buffer](https://github.com/echasnovski/mini.nvim/blob/334a2d9cfbd86ea08c4d870e39b5ba8bb354d938/doc/mini-pick.txt#L3540 (i.e. the buffer *in which* items are expected to be shown). So setting filetype in that buffer is done with `vim.bo[buf_id].filetype = ft`.

However, setting filetype usually has some non-trivial side effects. I'd suggest digging a little bit deeper and set highlighting directly. Take a look at [how default preview deals with this](https://github.com/echasnovski/mini.nvim/blob/334a2d9cfbd86ea08c4d870e39b5ba8bb354d938/lua/mini/pick.lua#L3071-L3073).

**Reply by abeldekat** [reply]
Thanks!

I hope this is more correct:

```lua
Pick.registry.buffer_lines_current = function()
  local show = function(buf_id, items, query, opts)
    local buf_target = buf_id
    if items and #items > 0 then
      local buf_src = items[1].bufnr -- all items from the same buffer
      local ft = vim.filetype.match({ buf = buf_src })

      local has_lang, lang = pcall(vim.treesitter.language.get_lang, ft)
      local has_ts, _ = pcall(vim.treesitter.start, buf_target, has_lang and lang or ft)
      if not has_ts and ft then vim.bo[buf_target].syntax = ft end
    end
    Pick.default_show(buf_target, items, query, opts)
  end
  MiniExtra.pickers.buf_lines({ scope = "current" }, { source = { show = show } })
end

```

**Reply by echasnovski** [reply]
I mean... The initial iteration was already correct, just less robust and efficient :)
And I still think that `local ft = vim.bo[items[1].bufnr].filetype` approach should be enough.

**Reply by abeldekat** [reply]
> And I still think that local ft = vim.bo[items[1].bufnr].filetype approach should be enough.

Thanks for the feedback! I also updated my initial code fragment, using the safer approach you suggested.

**Comment by pkazmier** [comment]
I’m very excited to try this later as it was something that I was going to look into for myself at some point!

**Comment by abeldekat** [comment]
![1719309644](https://github.com/echasnovski/mini.nvim/assets/58370433/6843aac2-7435-4c4f-8a7f-f7c2dc8278b1)


I *think* I now have a version where the highlighting is complete.
The idea: Override `vim.api.nvim_buf_set_lines`, to remove the linenumbers added to the buffer of mini.pick. Then, add the linenumbers as extmarks.

<details>

<summary>The code:</summary>

```lua
local Pick = require("mini.pick")
local H = {} -- Helper functions for custom pickers

---@type table<string,function>  event MiniPickStart
H.start_hooks = {}
---@type table<string,function> event MiniPickStop
H.stop_hooks = {}

H.setup_autocommands = function()
  local group = vim.api.nvim_create_augroup("minipick-hooks", { clear = true })
  local function au(pattern, desc, hooks)
    vim.api.nvim_create_autocmd({ "User" }, {
      pattern = pattern,
      group = group,
      desc = desc,
      callback = function(...)
        local opts = MiniPick.get_picker_opts() or {}
        if opts and opts.source then
          local hook = hooks[opts.source.name] or function(...) end
          hook(...)
        end
      end,
    })
  end
  au("MiniPickStart", "Picker start hook for source.name", H.start_hooks)
  au("MiniPickStop", "Picker stop hook for source.name", H.stop_hooks)
end

Pick.registry.buffer_lines_current = function()
  local ns = vim.api.nvim_create_namespace("ak")
  local show = function(buf_id, items, query, opts)
    local find_linenr = function(text)
      local _, _, linenr, _ = string.find(text, "(%d+):(.*)")
      return linenr
    end
    local remove_linenrs = function(replacement)
      return vim.tbl_map(function(item)
        local linenr = find_linenr(item)
        if linenr then item = string.sub(item, #linenr + 2) end
        return item
      end, replacement)
    end
    local add_extmark = function(buffer, idx, text)
      vim.api.nvim_buf_set_extmark(buffer, ns, idx - 1, 0, {
        virt_text = { { text } },
        virt_text_pos = "inline",
        priority = 1000, -- higher than treesitter's
      })
    end
    if not items or #items == 0 then Pick.default_show(buf_id, items, query, opts) end

    local set_lines_orig = vim.api.nvim_buf_set_lines
    ---@diagnostic disable-next-line: duplicate-set-field
    vim.api.nvim_buf_set_lines = function(buffer, start, end_, strict_indexing, replacement)
      set_lines_orig(buffer, start, end_, strict_indexing, remove_linenrs(replacement))
      for idx, _ in ipairs(replacement) do
        local linenr = find_linenr(replacement[idx])
        if linenr then add_extmark(buffer, idx, linenr .. ": ") end
      end
    end
    Pick.default_show(buf_id, items, query, opts)
    vim.api.nvim_buf_set_lines = set_lines_orig -- restore original function
  end

  local name = "Buffer lines (current)"
  local src_buffer_filetype = vim.bo.filetype -- remember ft of buffer to search in
  H.start_hooks[name] = function() vim.bo.filetype = src_buffer_filetype end -- set same ft in picker buffer
  H.stop_hooks[name] = function() vim.api.nvim_buf_clear_namespace(0, ns, 0, -1) end -- clean up
  MiniExtra.pickers.buf_lines({ scope = "current" }, { source = { show = show } })
end

H.setup_autocommands()

```

</details>

**EDIT 20240820**: [Latest version](https://github.com/echasnovski/mini.nvim/discussions/988#discussioncomment-10398788)

**Reply by echasnovski** [reply]
This looks... impressive. The idea to move line numbers into inline extmarks is a very nice touch.

However, if solution requires monkey-patching `vim.api.nvim_buf_set_lines()`, then there is something wrong either with solution itself or underlying design of source :)

Here is how I'd rewrite this:

```lua
local ns_digit_prefix = vim.api.nvim_create_namespace('cur-buf-pick-show')
local show_cur_buf_lines = function(buf_id, items, query, opts)
  if items == nil or #items == 0 then return end

  -- Show as usual
  MiniPick.default_show(buf_id, items, query, opts)

  -- Move prefix line numbers into inline extmarks
  local lines = vim.api.nvim_buf_get_lines(buf_id, 0, -1, false)
  local digit_prefixes = {}
  for i, l in ipairs(lines) do
    local _, prefix_end, prefix = l:find('^(%d+:)')
    if prefix_end ~= nil then
      digit_prefixes[i], lines[i] = prefix, l:sub(prefix_end + 1)
    end
  end

  vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, lines)
  for i, pref in pairs(digit_prefixes) do
    local opts = { virt_text = { { pref, 'MiniPickNormal' } }, virt_text_pos = 'inline' }
    vim.api.nvim_buf_set_extmark(buf_id, ns_digit_prefix, i - 1, 0, opts)
  end

  -- Set highlighting based on the curent filetype
  local ft = vim.bo[items[1].bufnr].filetype
  local has_lang, lang = pcall(vim.treesitter.language.get_lang, ft)
  local has_ts, _ = pcall(vim.treesitter.start, buf_id, has_lang and lang or ft)
  if not has_ts and ft then vim.bo[buf_id].syntax = ft end
end

MiniPick.registry.buffer_lines_current = function()
  MiniExtra.pickers.buf_lines({ scope = 'current' }, { source = { show = show_cur_buf_lines } })
end
```

This seems to work. However, filtering might rearrange line order which parser still might not handle well. So as much as I like the idea of highlighted "current buffer lines" picker, I don't think it has an easy solution under current design. One idea might be to highlight only once, somehow get and save how highlighting is done, disable tree-sitter highlighting, and then re-higlight manually. But this is too close to being magic.

**Reply by abeldekat** [reply]
Indeed: monkey-patching `vim.api.nvim_buf_set_lines()`... I can't even convince myself that this is a prime example of the decorator pattern...)). 

The benefit of the patch: Lines are added only once to the buffer during a "show" iteration. Also, the letters typed when searching are still highlighted. However, I now see that the highlighting in a word fragment is incorrect by a certain index, which renders this advantage null and void...

During filtering, [telescope's](https://github.com/nvim-telescope/telescope.nvim/blob/61a4a615366c470a4e9ca8f8b45718b6b92af73f/lua/telescope/builtin/__files.lua#L470) highlight results are better. The algorithm is much more complicated though. I only try to add code to my config that I can understand somewhat easily...

I think the solution is "good enough" for me. Even during filtering, the highlighting "holds" on a acceptable level. 

I will use your code, as it's more concise and readable. Thanks!

ps: I watched the mini.git demo yesterday and enjoyed seeing `Neo McVim` in action.

**Reply by echasnovski** [reply]
> ps: I watched the mini.git demo yesterday and enjoyed seeing `Neo McVim` in action.

Yeah, he's the *real* Git guru :)

**Reply by pkazmier** [reply]
> ps: I watched the mini.git demo yesterday and enjoyed seeing `Neo McVim` in action.

Link?

**Reply by echasnovski** [reply]
> > ps: I watched the mini.git demo yesterday and enjoyed seeing `Neo McVim` in action.
> 
> Link?

https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-git.md#demo

**Reply by pkazmier** [reply]
Doh! You can tell I had just woke up as I thought someone had done a demo on YouTube that I hadn’t seen, so I was googling “Neo McVim” and was coming up empty handed! 😀

**Reply by abeldekat** [reply]
It's somewhat reassuring that google still does not know everything...)

**Reply by abeldekat** [reply]
I now have doubts about the "extmarks" approach. The results are not correct(too many) when adding a modifier like `'` before the query. I compared with the original `Pick buf_lines scope='current'`.

I assumed that `extmarks` are not taken into account by the search algorithm. Perhaps it's better to use the first version without extmarks.

**Reply by abeldekat** [reply]
Somehow, the problem above can be fixed by positioning the linenumbers in the extmark on the right:

```lua
    -- ...
    for i, pref in pairs(digit_prefixes) do
      pref = string.sub(pref, 1, -2)
      local extmark_opts = { virt_text = { { pref } }, virt_text_pos = "right_align" }
      vim.api.nvim_buf_set_extmark(buf_id, ns_digit_prefix, i - 1, 0, extmark_opts)
    end
    -- ...
```

**Reply by lsig-ks** [reply]
I am trying this and it works great, but I was wondering if it is possible to sort the output of the lines in ascending order (1, 2, 3, 4). When I try it not the order seems random.

**Reply by echasnovski** [reply]
> I am trying this and it works great, but I was wondering if it is possible to sort the output of the lines in ascending order (1, 2, 3, 4). When I try it not the order seems random.

This is default matching approach: it sorts based on the [following criteria](https://github.com/echasnovski/mini.nvim/blob/23669ab7abc2cd10f9a5166de07400bb34eaecda/doc/mini-pick.txt#L871-L873). I've indeed encountered several times where preserving order might be a good idea, so thinking about the better non-blocking approach to allow this.

**Reply by lsig-ks** [reply]
Great, I look forward to see what you come up with. In the mean time Ill keep using this approach, which is in most cases good enough. 

P.S. Thanks for the great work!

**Reply by echasnovski** [reply]
@lsig-ks, `MiniExtra.pickers.buf_lines` now (on latest `main`) has `preserve_order` local option which if `true` will preserve original order of lines. You can use it like `:Pick buf_lines scope='current' preserve_order=true`. This will use non-blocking matching following the [change to `MiniPick.default_match()`](https://github.com/echasnovski/mini.nvim/commit/f045e9d89608064fe5b01fd5f7333d5f5012f17a).

**Reply by pkazmier** [reply]
Excited to test this weekend when I'm home from my business trip.

**Reply by pkazmier** [reply]
This works great. The only changes I made above to @echasnovski's snippet above was to use the new option he made and to force an exact match as the fuzzy match is not as useful when order is preserved, so I changed these lines:
```lua
MiniPick.registry.buffer_lines_current = function()
  MiniExtra.pickers.buf_lines({ scope = 'current' }, { source = { show = show_cur_buf_lines } })
end
```

to these lines:
```lua
MiniPick.registry.buffer_lines_current = function()
  MiniExtra.pickers.buf_lines({ scope = "current", preserve_order = true }, {
    source = {
      match = function(stritems, inds, query, opts)
        table.insert(query, 1, "'")
        return MiniPick.default_match(stritems, inds, query, opts)
      end,
      show = show_cur_buf_lines,
    },
  })
end
```

Now, I'm going to change my `<leader>/` binding to this new picker instead of my old binding `Pick buf_lines scope='current'`. Very excited!

**Reply by pkazmier** [reply]
Hmm ... I think I need a different matching function. Exact is good for a single word, but then you need to `<C-Space>` to refine the query to search for the next word. I want a matching function that isn't fuzzy, but lets me choose terms separated by whitespace. So, "hello world" would match this line "hello Pete, this is the world speaking to you". Going back to the docs to see if there is something that would work.

**Reply by echasnovski** [reply]
> So, \"hello world\" would match this line \"hello Pete, this is the world speaking to you\". Going back to the docs to see if there is something that would work.

This is how default matching works. I'd suggest having an explicit exact match by adding a space afterwards. This does not require thinking ahead of time whether fuzzy or exact matching is better.

**Reply by pkazmier** [reply]
Oh yeah, I was overthinking it! Thanks!

**Reply by pkazmier** [reply]
I had to make a small tweak to @echasnovski's sample code as the default show function doesn't use `:` but rather `│`. I didn't realize it was not working as intended until I tried to format the number column. So, here is a working version for anyone that comes along afterwards:

```lua
local ns_digit_prefix = vim.api.nvim_create_namespace("cur-buf-pick-show")
local show_cur_buf_lines = function(buf_id, items, query, opts)
  if items == nil or #items == 0 then
    return
  end

  -- Show as usual
  MiniPick.default_show(buf_id, items, query, opts)

  -- Move prefix line numbers into inline extmarks
  local lines = vim.api.nvim_buf_get_lines(buf_id, 0, -1, false)
  local digit_prefixes = {}
  for i, l in ipairs(lines) do
    local _, prefix_end, prefix = l:find("^(%d+│)")
    if prefix_end ~= nil then
      digit_prefixes[i], lines[i] = prefix, l:sub(prefix_end + 1)
    end
  end

  vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, lines)
  for i, pref in pairs(digit_prefixes) do
    local opts = { virt_text = { { string.format("%8.8s", pref), "MiniPickNormal" } }, virt_text_pos = "inline" }
    vim.api.nvim_buf_set_extmark(buf_id, ns_digit_prefix, i - 1, 0, opts)
  end

  -- Set highlighting based on the curent filetype
  local ft = vim.bo[items[1].bufnr].filetype
  local has_lang, lang = pcall(vim.treesitter.language.get_lang, ft)
  local has_ts, _ = pcall(vim.treesitter.start, buf_id, has_lang and lang or ft)
  if not has_ts and ft then
    vim.bo[buf_id].syntax = ft
  end
end

MiniPick.registry.buffer_lines_current = function()
  MiniExtra.pickers.buf_lines({ scope = "current", preserve_order = true }, { source = { show = show_cur_buf_lines } })
end
```

That results in line numbers being aligned at the expense of a little wasted horizontal space, but I think it looks much nicer:

https://github.com/user-attachments/assets/fb5ece7e-4696-4c7a-8ffa-e11afda4b169

**Reply by echasnovski** [reply]
> That results in line numbers being aligned at the expense of a little wasted horizontal space, but I think it looks much nicer:

You know what? This indeed looks much better. I think I'll make this default in `MiniExtra.pickers.buf_lines`. And quite a concise change to the code too.

**Reply by abeldekat** [reply]
> I had to make a small tweak to @echasnovski's sample code as the default show function doesn't use : but rather │.

My apologies. I fixed this some time ago but forgot to update the code fragment in this discussion.

> That results in line numbers being aligned at the expense of a little wasted horizontal space, but I think it looks much nicer:

Indeed! Excited to try this.
Edit: Included in my config, thanks!

**Reply by pkazmier** [reply]
@abeldekat I'm so glad your pursued this and had the idea of using extmarks to allow the highlighting to work correctly—very clever! I'm loving the way this picker looks now!

**Reply by abeldekat** [reply]
Thanks! I had a lot of fun pursuing...;)

**Reply by echasnovski** [reply]
> > That results in line numbers being aligned at the expense of a little wasted horizontal space, but I think it looks much nicer:
> 
> You know what? This indeed looks much better. I think I'll make this default in `MiniExtra.pickers.buf_lines`. And quite a concise change to the code too.

@abeldekat, @pkazmier, `MiniExtra.pickers.buf_lines` now pads line numbers (since https://github.com/echasnovski/mini.nvim/commit/0b0cc07697e298351b6543d5d4db5372b71897f2). Thanks for the demo!

This also means that the code for this "colored buffer lines" picker needs to be updated. Sorry :shrug:

**Reply by abeldekat** [reply]
That's great!

@echasnovski, @pkazmier, I propose the following "latest version"

```lua
local ns_digit_prefix = vim.api.nvim_create_namespace("cur-buf-pick-show")
local show_cur_buf_lines = function(buf_id, items, query, opts)
  if items == nil or #items == 0 then
    return
  end

  -- Show as usual
  MiniPick.default_show(buf_id, items, query, opts)

  -- Move prefix line numbers into inline extmarks
  local lines = vim.api.nvim_buf_get_lines(buf_id, 0, -1, false)
  local digit_prefixes = {}
  for i, l in ipairs(lines) do
    local _, prefix_end, prefix = l:find("^(%s*%d+│)")
    if prefix_end ~= nil then
      digit_prefixes[i], lines[i] = prefix, l:sub(prefix_end + 1)
    end
  end

  vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, lines)
  for i, pref in pairs(digit_prefixes) do
    local opts = { virt_text = { { pref, "MiniPickNormal" } }, virt_text_pos = "inline" }
    vim.api.nvim_buf_set_extmark(buf_id, ns_digit_prefix, i - 1, 0, opts)
  end

  -- Set highlighting based on the curent filetype
  local ft = vim.bo[items[1].bufnr].filetype
  local has_lang, lang = pcall(vim.treesitter.language.get_lang, ft)
  local has_ts, _ = pcall(vim.treesitter.start, buf_id, has_lang and lang or ft)
  if not has_ts and ft then vim.bo[buf_id].syntax = ft end
end

MiniPick.registry.buffer_lines_current = function()
  -- local local_opts = { scope = "current", preserve_order = true } -- use preserve_order
  local local_opts = { scope = "current" }
  MiniExtra.pickers.buf_lines(local_opts, { source = { show = show_cur_buf_lines } })
end
```

Perhaps its better to use the default settings, and add `preserve_order` as a suggestion.
I will add links to the latest version to the top of the post, and to the start of this thread.

**Reply by pkazmier** [reply]
Thanks. I've updated my config with the new version above (except I prefer preserving order as those out of order numbers on the left make my brain hurt 😄 )
